docs/configuration.md
# Configuration
* [Overview][0]
* [Data configuration][1]
  * [Sample JSON][2]
  * [Parameter reference][3]
* [Plugin configuration][4]
* [Kubernetes component definitions][5]
  * [Overview][6]
  * [RBAC][7]
  * [Pod][8]

## Overview

Sonobuoy's configuration can be split into three conceptual parts: (1) data configuration (2) plugin configuration (3) Kubernetes component definitions. Their file format depends on how Sonobuoy is run:
* **Standalone**: JSON files that live on a cluster node
* **Containerized**: YAML manifests that are pushed to the API Server using `kubectl`

Once the configs are loaded (in either case), Sonobuoy parses them and gathers data accordingly. If plugins are specified, Sonobuoy [submits worker pods to collect node-specific information][9].


| | Overview|Path on Cluster Node|[STANDALONE]<br>JSON example(s)|[CONTAINERIZED]<br>YAML manifest example(s)
|---|---|---|---|---|
|*Data configuration*| What Sonobuoy records, how, and where. |*ANY of the following*:<br>(1) `config.json` in the directory where `sonobuoy` is executed<br>(2) `/etc/sonobuoy/config.json`<br>(3) `$SONOBUOY_CONFIG`<br><br>|[`config.json`][10]|<br> [`examples/ksonnet/components/10-configmaps.jsonnet`][11]<br><br>*The jsonnet file is basically a wrapper for the `config.json` file, which allows it to be properly mounted onto the cluster's Sonobuoy pod.* <br><br>
|*Plugin configuration*|Settings for each plugin integration.|*ANY of the following*:<br>(1) `/etc/sonobuoy/plugins.d`<br>(2) `$HOME/.sonobuoy/plugins.d`<br>(3) `./plugins.d`<br>(4) `PluginSearchPath` (override from the data configuration) <br><br>| There is a YAML config for each plugin:<br>(1) [`plugins.d/e2e.yaml`][16]<br>(2)[`plugins.d/systemdlogs.yaml`][17]|<br>[`examples/ksonnet/components/10-configmaps.jsonnet`][11]<br><br>*Same comment about the jsonnet file as above.*
|*Kubernetes component definitions*|The various K8s objects that need to be defined for Sonobuoy to run as a containerized app.|N/A (manifest only)|N/A|The example splits this into two manifests:<br>(1) [`examples/ksonnet/components/00-rbac.jsonnet`][12]<br>(2) [`examples/ksonnet/components/20-pod.jsonnet`][13]|



*NOTE: All of these specifications have been consolidated into one YAML file (as in [`examples/quickstart.yaml`][18]), as long as any RBAC settings are defined first. *


## Data configuration

### Sample JSON
See [Parameter Reference][3] for a more detailed description of each setting.
```json
{
    "UUID": "12345-something-unique",
    "Description": "EXAMPLE",
    "Version": "v0.3.0",
    "Kubeconfig": "~/.kube/config",
    "ResultsDir": "./results",
    "Resources": [
        "CertificateSigningRequests",
        "ClusterRoleBindings",
        "ClusterRoles",
        "ComponentStatuses",
        "Nodes",
        "PersistentVolumes",
        "PodSecurityPolicies",
        "ServerVersion",
        "StorageClasses",
        "ThirdPartyResources",
        "ConfigMaps",
        "DaemonSets",
        "Deployments",
        "Endpoints",
        "Events",
        "HorizontalPodAutoscalers",
        "Ingresses",
        "Jobs",
        "LimitRanges",
        "PersistentVolumeClaims",
        "Pods",
        "PodLogs",
        "PodDisruptionBudgets",
        "PodPresets",
        "PodTemplates",
        "ReplicaSets",
        "ReplicationControllers",
        "ResourceQuotas",
        "RoleBindings",
        "Roles",
        "Secrets",
        "ServerGroups",
        "ServiceAccounts",
        "Services",
        "StatefulSets"
    ],
    "Filters": {
        "LabelSelector": "",
        "Namespaces": ".*"
    },
    "Limits": {
        "PodLogs": {
            "LimitTime": "24h"
        }
    },
    "Server": {
        "advertiseaddress": "",
        "bindaddress": "0.0.0.0",
        "bindport": 8080,
        "timeoutseconds": 5400
    },
    "Plugins": [
        {"name": "e2e"},
        {"name": "systemd_logs"}
    ],
    "PluginSearchPath": "override-path"
}
```
### Parameter reference
| Key  | Type | Default | Meaning  |
|---|---|---|---|
|  UUID | String |  A randomly generated UUID | Used to name the directory where Sonobuoy-collected data is saved (e.g. `sonobuoy_<UUID>`)  |
| Description | String | "DEFAULT" | Metadata for keeping track of configs and their use cases |
| Version | String | `buildInfo.version` (comes from the compilation of the Docker image) | The Sonobuoy version that the config uses for its schema |
| Kubeconfig | String | ""<br><br>*NOT necessary for containerized Sonobuoy, which will default to the in-cluster config* | Allows Sonobuoy to communicate with and gather info from the Kubernetes cluster |
| ResultsDir | String | "./results" | The directory in which Sonobuoy writes its results. Customizable with `RESULTS_DIR` environment variable. |
| Resources | String Array | An array containing all possible resources | *See the [sample JSON][2] above for a list of all available resource types.*<br><br>Indicates to Sonobuoy what type of data it should be recording |
| Filters.LabelSelector | String | "" | Uses standard Kubernetes [label selector syntax][14] to filter which resource objects are recorded |
| Filters.Namespaces | String | ".*" | Uses regex on namespaces to filter which resource objects are recorded |
| Limits.PodLogs.LimitTime | String | "" | Limits how far back in time to gather Pod Logs, leave blank for no limit (e.g. "24h", "60m". See https://golang.org/pkg/time/#ParseDuration for details.) |
| Limits.PodLogs.LimitSize | String | "" | Limits the size of Pod Logs to gather, per container, leave blank for no limit (e.g. "10 MB", "1 GB", etc.) |
| Server.advertiseaddress | String | `$SONOBUOY_ADVERTISE_IP` &#124;&#124; the current server's `os.Hostname()`| *Only used if Sonobuoy dispatches agent pods to collect node-specific information*<br><br>The IP address that remote Sonobuoy agents send information back to, in order for disparate data to be aggregated into a single report |
| Server.bindaddress | String | "0.0.0.0" | *See `Server.advertiseaddress` for context.*<br><br>If data aggregation is required, an HTTP server is started to handle the worker requests. This is the address that server binds to. |
| Server.bindport | Int | 8080 | The port for the HTTP server mentioned in *Server.bindaddress*. |
| Server.timeoutseconds | Int | 5400 (90 min) | *See `Server.advertiseaddress` for context.*<br><br>This determines how long the master Sonobuoy pod should wait to hear back from the dispatched agents. |
| Plugins | Array of plugin descriptions: `{"name": <PLUGIN_NAME>}` | `[]` | The list of Sonobuoy plugins enabled for custom data collection. See the [plugins reference][9] for details.|
| PluginSearchPath | String Array | `"./plugins.d", "/etc/sonobuoy/plugins.d", "~/sonobuoy/plugins.d"` | The paths where Sonobuoy should look for its plugin configs

## Plugin configuration

If you have the appropriate plugin definition (third-party or your own), you can easily opt into it by adding a *Plugin Selection* to the main Sonobuoy config's "Plugin" section.

The following code snippet, for example, selects the `e2e` and `systemd_logs` plugins:

```
    "Plugins": [
        {"name": "e2e"},
        {"name": "systemd_logs"}
    ]
```

For more details on creating custom plugins, see the [plugin reference][9].

## Kubernetes component definitions

### Overview

*This section of the configuration is only applicable when running Sonobuoy as a containerized pod.*

While the other configuration sections control Sonobuoy's data collection, this part defines the Kubernetes resources required to actually run Sonobuoy on your cluster. In the [`examples/ksonnet/components`][19] example, it is split into two manifests:
1. `00-rbac.jsonnet`: This sets up gives Sonobuoy the necessary permissions to query the API server.
2. `20-pod.jsonnet`: This sets up the Sonobuoy Pod and associated Service.

### RBAC

If you are using the provided manifests as a basis for your own, *you will not need to change this file.* The ServiceAccount, ClusterRoleBinding, and ClusterRole declarations should be left as they are.


### Pod



However, ensure that your pod declaration has addressed the following aspects, which may be unique to your cluster:
* **Image Name**
  * The provided example configs pull from Heptio's registry of pre-built Sonobuoy container images (*gcr.io/heptio-images/sonobuoy*).
  * To use custom-built images from your private registry, make sure to change the `image` and `imagePullPolicy` config fields.


* **Config Volume Mount**
  * To mount the main Sonobuoy config on the Sonobuoy pod, a ConfigMap is recommended. This is used in the example.
  * Ensure that `mountPath` matches one of the paths enumerated in [the chart above][0].


* **Output Volume Mount**
  * Sonobuoy output should be persisted by writing to [Persistent Volume Claims (PVC)][15].
  * However, for the sake of prototyping (as in the example), you can mount via `emptyDir`. This allows you to later inspect Sonobuoy's results by copying them to your local machine with `kubectl cp`.


* **Advertise IP (env variable)**
  * It's recommended to keep this the same as in the example config, where Kubernetes auto-populates the environmental variable `$SONOBUOY_ADVERTISE_IP` with the IP of the assigned pod.

[0]: #Overview
[1]: #data-configuration
[2]: #sample-json
[3]: #parameter-reference
[4]: #plugin-configuration
[5]: #kubernetes-component-definitions
[6]: #overview-1
[7]: #rbac
[8]: #pod
[9]: plugins.md
[10]: /config.json
[11]: /examples/ksonnet/components/10-configmaps.jsonnet
[12]: /examples/ksonnet/components/00-rbac.jsonnet
[13]: /examples/ksonnet/components/20-pod.jsonnet
[14]: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
[15]: https://kubernetes.io/docs/tasks/configure-pod-container/configure-persistent-volume-storage/
[16]: /plugins.d/e2e.yaml
[17]: plugins.d/systemdlogs.yaml
[18]: /examples/quickstart.yaml
[19]: /examples/ksonnet/components


CHANGELOG.md
# Changelog

#### v0.8.0 - 2017-08-03
  * Initial Release


pkg/config/config.go
/*
Copyright 2017 Heptio Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package config

import (
	"path"
	"time"

	"github.com/c2h5oh/datasize"
	"github.com/heptio/sonobuoy/pkg/buildinfo"
	"github.com/heptio/sonobuoy/pkg/plugin"
	"github.com/satori/go.uuid"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

///////////////////////////////////////////////////////
// Note: The described resources are a 1:1 match
// with kubectl UX for consistent user experience.
// xref: https://kubernetes.io/docs/api-reference/v1.8/
///////////////////////////////////////////////////////

// ClusterResources is the list of API resources that are scoped to the entire
// cluster (ie. not to any particular namespace)
var ClusterResources = []string{
	"CertificateSigningRequests",
	"ClusterRoleBindings",
	"ClusterRoles",
	"ComponentStatuses",
	"CustomResourceDefinitions",
	"Nodes",
	"PersistentVolumes",
	"PodSecurityPolicies",
	"StorageClasses",
	"ThirdPartyResources",
}

// NamespacedResources is the list of API resources that are scoped to a
// kubernetes namespace.
var NamespacedResources = []string{
	"ConfigMaps",
	"ControllerRevisions",
	"CronJobs",
	"DaemonSets",
	"Deployments",
	"Endpoints",
	"Events",
	"HorizontalPodAutoscalers",
	"Ingresses",
	"Jobs",
	"LimitRanges",
	"NetworkPolicies",
	"PersistentVolumeClaims",
	"PodDisruptionBudgets",
	"PodPresets",
	"PodTemplates",
	"Pods",
	"ReplicaSets",
	"ReplicationControllers",
	"ResourceQuotas",
	"RoleBindings",
	"Roles",
	"Secrets",
	"ServiceAccounts",
	"Services",
	"StatefulSets",
}

// SpecialResources are resources that aren't queried (or stored) the same was
// as the rest, so need special casing for querying them.
var SpecialResources = []string{
	"PodLogs",
	"ServerGroups",
	"ServerVersion",
}

// FilterOptions allow operators to select sets to include in a report
type FilterOptions struct {
	Namespaces    string `json:"Namespaces"`
	LabelSelector string `json:"LabelSelector"`
}

// Config is the input struct used to determine what data to collect.
type Config struct {
	// NOTE: viper uses "mapstructure" as the tag for config
	// serialization, *NOT* "json".  mapstructure is a separate library
	// that converts maps to structs, and has its own syntax for tagging
	// fields. The only documentation on this is in the mapstructure docs:
	// https://godoc.org/github.com/mitchellh/mapstructure#example-Decode--Tags
	//
	// To be safe we annotate with both json and mapstructure tags.

	///////////////////////////////////////////////
	// Meta-Data collection options
	///////////////////////////////////////////////
	Description string `json:"Description" mapstructure:"Description"`
	UUID        string `json:"UUID" mapstructure:"UUID"`
	Version     string `json:"Version" mapstructure:"Version"`
	ResultsDir  string `json:"ResultsDir" mapstructure:"ResultsDir"`
	Kubeconfig  string `json:"Kubeconfig" mapstructure:"Kubeconfig"`

	///////////////////////////////////////////////
	// Data collection options
	///////////////////////////////////////////////
	Resources []string `json:"Resources" mapstructure:"Resources"`

	///////////////////////////////////////////////
	// Filtering options
	///////////////////////////////////////////////
	Filters FilterOptions `json:"Filters" mapstructure:"Filters"`

	///////////////////////////////////////////////
	// Limit options
	///////////////////////////////////////////////
	Limits LimitConfig `json:"Limits" mapstructure:"Limits"`

	///////////////////////////////////////////////
	// plugin configurations settings
	///////////////////////////////////////////////
	Aggregation      plugin.AggregationConfig `json:"Server" mapstructure:"Server"`
	PluginSelections []plugin.Selection       `json:"Plugins" mapstructure:"Plugins"`
	PluginSearchPath []string                 `json:"PluginSearchPath" mapstructure:"PluginSearchPath"`
	PluginNamespace  string                   `json:"PluginNamespace" mapstructure:"PluginNamespace"`
	LoadedPlugins    []plugin.Interface       // this is assigned when plugins are loaded.
}

// LimitConfig is a configuration on the limits of sizes of various responses.
type LimitConfig struct {
	PodLogs SizeOrTimeLimitConfig `json:"PodLogs" mapstructure:"PodLogs"`
}

// SizeOrTimeLimitConfig represents configuration that limits the size of
// something either by a total disk size, or by a length of time.
type SizeOrTimeLimitConfig struct {
	LimitSize string `json:"LimitSize" mapstructure:"LimitSize"`
	LimitTime string `json:"LimitTime" mapstructure:"LimitTime"`
}

// FilterResources is a utility function used to parse Resources
func (cfg *Config) FilterResources(filter []string) map[string]bool {
	results := make(map[string]bool)

	for _, felement := range filter {
		for _, check := range cfg.Resources {
			if felement == check {
				results[felement] = true
			}
		}
	}
	return results
}

// OutputDir returns the directory under the ResultsDir containing the
// UUID for this run.
func (cfg *Config) OutputDir() string {
	return path.Join(cfg.ResultsDir, cfg.UUID)
}

// SizeLimitBytes returns how many bytes the configuration is set to limit,
// returning defaultVal if not set.
func (c SizeOrTimeLimitConfig) SizeLimitBytes(defaultVal int64) int64 {
	val, defaulted, err := c.sizeLimitBytes()

	// Ignore error, since we should have already caught it in validation
	if err != nil || defaulted {
		return defaultVal
	}

	return val
}

func (c SizeOrTimeLimitConfig) sizeLimitBytes() (val int64, defaulted bool, err error) {
	str := c.LimitSize
	if str == "" {
		return 0, true, nil
	}

	var bs datasize.ByteSize
	err = bs.UnmarshalText([]byte(str))
	return int64(bs.Bytes()), false, err
}

// TimeLimitDuration returns the duration the configuration is set to limit, returning defaultVal if not set.
func (c SizeOrTimeLimitConfig) TimeLimitDuration(defaultVal time.Duration) time.Duration {
	val, defaulted, err := c.timeLimitDuration()

	// Ignore error, since we should have already caught it in validation
	if err != nil || defaulted {
		return defaultVal
	}

	return val
}

func (c SizeOrTimeLimitConfig) timeLimitDuration() (val time.Duration, defaulted bool, err error) {
	str := c.LimitTime
	if str == "" {
		return 0, true, nil
	}

	val, err = time.ParseDuration(str)
	return val, false, err
}

// NewWithDefaults returns a newly-constructed Config object with default values.
func NewWithDefaults() *Config {
	var cfg Config
	cfg.UUID = uuid.NewV4().String()
	cfg.Description = "DEFAULT"
	cfg.ResultsDir = "./results"
	cfg.Version = buildinfo.Version

	cfg.Filters.Namespaces = ".*"

	cfg.Resources = ClusterResources
	cfg.Resources = append(cfg.Resources, NamespacedResources...)
	cfg.Resources = append(cfg.Resources, SpecialResources...)

	cfg.PluginNamespace = metav1.NamespaceSystem

	cfg.Aggregation.BindAddress = "0.0.0.0"
	cfg.Aggregation.BindPort = 8080
	cfg.Aggregation.TimeoutSeconds = 5400 // 90 minutes

	cfg.PluginSearchPath = []string{
		"./plugins.d",
		"/etc/sonobuoy/plugins.d",
		"~/sonobuoy/plugins.d",
	}

	return &cfg
}

// addPlugin adds a (configured, initialized) plugin to the config object so
// that it can be executed.
func (cfg *Config) addPlugin(plugin plugin.Interface) {
	cfg.LoadedPlugins = append(cfg.LoadedPlugins, plugin)
}

// getPlugins gets the list of plugins selected for this configuration.
func (cfg *Config) getPlugins() []plugin.Interface {
	return cfg.LoadedPlugins
}
